code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/deployment/impl/DeploymentManagerImpl.java

```diff
@@ -0,0 +1,364 @@
+                Map<Service, DeploymentUnitInstance> serviceToInstanceMap = new HashMap<>();
+                if (uuidToInstances.get(deploymentUnitInstance.getUuid()) != null) {
+                    serviceToInstanceMap = uuidToInstances.get(deploymentUnitInstance.getUuid());
+                }
+                serviceToInstanceMap.put(deploymentUnitInstance.getService(), deploymentUnitInstance);
+                uuidToInstances.put(deploymentUnitInstance.getUuid(), serviceToInstanceMap);
+            }
+        }
+        for (String uuid : uuidToInstances.keySet()) {
+            deploymentUnits.add(new DeploymentUnit(uuid, uuidToInstances.get(uuid)));
+        }
+
+        return deploymentUnits;
+    }
+
+    protected void activateDeploymentUnits(List<Service> services, List<DeploymentUnit> units, int requestedScale) {
+        /*
+         * Delete invalid units
+         */
+        units = deleteBadUnits(units);
+
+        /*
+         * Create new units / remove old units to match the scale
+         */
+        units = matchScale(services, units, requestedScale);
+
+        /*
+         * Fill up the gaps in units
+         */
+        fillUpMissingUnitInstances(services, units);
+
+        /*
+         * Activate all the units
+         */
+        startUnits(units);
+
+        /*
+         * Delete the units that have a bad health
+         */
+
+        cleanup(units);
+    }
+
+    protected void cleanup(List<DeploymentUnit> units) {
+        for (DeploymentUnit unit : units) {
+            if (unit.needsCleanup()) {
+                unit.cleanup();
+            }
+        }
+    }
+
+    protected void startUnits(List<DeploymentUnit> units) {
+        for (DeploymentUnit unit : units) {
+            unit.start();
+        }
+    }
+
+    protected List<DeploymentUnit> deleteBadUnits(List<DeploymentUnit> units) {
+        List<DeploymentUnit> result = new ArrayList<>(units.size());
+
+        for (DeploymentUnit unit : units) {
+            if (unit.isError()) {
+                unit.remove();
+            } else {
+                result.add(unit);
+            }
+        }
+
+        return result;
+    }
+
+    protected List<DeploymentUnit> matchScale(List<Service> services, List<DeploymentUnit> units, int requestedScale) {
+        /*
+         * If there are too many units, call unit.remove() and delete the excess.
+         * If there are not enough units, create new empty deployment units by call the factory
+         * and passing in the services and an empty list for instances.
+         * 
+         * NOTE: We don't actually start services hereed
+         */
+
+        if (units.size() < requestedScale) {
+            addMissingUnits(services, units, requestedScale);
+        } else if (units.size() > requestedScale) {
+            removeExtraUnits(units, requestedScale);
+        }
+
+        return units;
+    }
+
+    private void removeExtraUnits(List<DeploymentUnit> units, int scale) {
+        // delete units
```

Comment:
No, order doesn't matter neither between the units, nor within the units. I agree that we might optimize it in the future.